src/tools/w3dviewer/mainfrm.cpp

```diff
@@ -0,0 +1,1610 @@
+     ON_UPDATE_COMMAND_UI(ID_OBJECT_RESTRICTANIMS, OnUpdateRestrictAnims)
+     ON_COMMAND(ID_AGGREGATE_BINDSUBOBJECTLOD, OnBindSubobject)
+     ON_UPDATE_COMMAND_UI(ID_AGGREGATE_BINDSUBOBJECTLOD, OnUpdateBindSubobject)
+     ON_COMMAND(ID_CAMERA_SETDISTANCE, OnSetDistance)
+     ON_COMMAND(ID_OBJECT_TOGGLEALTERNATEMATERIALS, OnAlternateMaterials)
+     ON_COMMAND(ID_SOUND_CREATESOUNDOBJECT, OnCreateSoundObject)
+     ON_COMMAND(ID_SOUND_EDITSOUNDOBJECT, OnEditSoundObject)
+     ON_UPDATE_COMMAND_UI(ID_SOUND_EDITSOUNDOBJECT, OnUpdateEditSoundObject)
+     ON_COMMAND(ID_EXPORT_SOUNDOBJECT, OnExportSoundObject)
+     ON_UPDATE_COMMAND_UI(ID_EXPORT_SOUNDOBJECT, OnUpdateEditSoundObject)
+     ON_COMMAND(ID_VIEW_WIREFRAMEMODE, OnWireframeMode)
+     ON_UPDATE_COMMAND_UI(ID_VIEW_WIREFRAMEMODE, OnUpdateWireframeMode)
+     ON_UPDATE_COMMAND_UI(ID_BACKGROUND_FOG, OnUpdateFog)
+     ON_COMMAND(ID_BACKGROUND_FOG, OnFog)
+     ON_UPDATE_COMMAND_UI(ID_EMITTERS_SCALEEMITTER, OnUpdateScaleEmitter)
+     ON_COMMAND(ID_EMITTERS_SCALEEMITTER, OnScaleEmitter)
+     ON_UPDATE_COMMAND_UI(ID_VIEW_POLYGONSORTING, OnUpdatePolygonSorting)
+     ON_COMMAND(ID_VIEW_POLYGONSORTING, OnPolygonSorting)
+     ON_COMMAND(ID_CAMERA_XCAMERA, OnPlusXCamera)
+     ON_UPDATE_COMMAND_UI(ID_CAMERA_XCAMERA, OnUpdatePlusXCamera)
+     ON_COMMAND(ID_FILE_MUNGESORTONLOAD, OnMungeSort)
+     ON_UPDATE_COMMAND_UI(ID_FILE_MUNGESORTONLOAD, OnUpdateMungeSort)
+     ON_COMMAND(ID_FILE_ENABLEGAMMACORRECTION, OnEnableGamma)
+     ON_UPDATE_COMMAND_UI(ID_FILE_ENABLEGAMMACORRECTION, OnUpdateEnableGamma)
+     ON_COMMAND(ID_VIEW_SETGAMMA, OnSetGamma)
+END_MESSAGE_MAP()
+// clang-format on
+
+int DeviceWidth = 640;
+int DeviceHeight = 480;
+int Device = -1;
+int BPP = -1;
+
+CMainFrame::CMainFrame() :
+    m_currentType(ASSET_TYPE_NONE), m_animationToolbarVisible(true), m_initialized(false), m_subMenu(nullptr), m_windowRect{}
+{
+}
+
+CMainFrame::~CMainFrame() {}
+
+BOOL CMainFrame::PreCreateWindow(CREATESTRUCT &cs)
+{
+    return CFrameWnd::PreCreateWindow(cs);
+}
+
+BOOL CMainFrame::OnCmdMsg(UINT nID, int nCode, void *pExtra, AFX_CMDHANDLERINFO *pHandlerInfo)
+{
+    if (nCode != -1) {
+        return CFrameWnd::OnCmdMsg(nID, nCode, pExtra, pHandlerInfo);
+    }
+
+    if (pExtra == nullptr) {
+        return CFrameWnd::OnCmdMsg(nID, nCode, pExtra, pHandlerInfo);
+    }
+
+    CCmdUI *pCmdUI = static_cast<CCmdUI *>(pExtra);
+
+    if (pCmdUI->m_nID < 1000 || pCmdUI->m_nID > 1100) {
+        return CFrameWnd::OnCmdMsg(nID, nCode, pExtra, pHandlerInfo);
+    }
+
+    pCmdUI->Enable();
+    return TRUE;
+}
+
+BOOL CMainFrame::OnCommand(WPARAM wParam, LPARAM lParam)
+{
+    if (lParam == 0 && LOWORD(wParam) >= 1000 && LOWORD(wParam < 1100)) {
+        MENUITEMINFO info;
+        memset(&info, 0, sizeof(info));
+        info.cbSize = sizeof(info);
+        info.fMask = MIIM_DATA | MIIM_TYPE;
+        info.dwTypeData = reinterpret_cast<char *>(&info.hbmpItem);
```

Comment:
Did you check if this is legal? It looks very suspicious. A quick glance at documentation does not make it obvious if this is proper.

https://docs.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-menuiteminfoa

Reply:
Found out what was going on and fixed it.